grpc.go

```diff
@@ -0,0 +1,698 @@
+	}
+
+	n, err := file.Write(req.Data)
+	if err != nil {
+		return &pb.WriteStreamResponse{}, err
+	}
+
+	return &pb.WriteStreamResponse{
+		Len: uint32(n),
+	}, nil
+}
+
+func (a *agentGRPC) ReadStdout(ctx context.Context, req *pb.ReadStreamRequest) (*pb.ReadStreamResponse, error) {
+	data, err := a.pod.readStdio(req.ContainerId, req.ProcessId, int(req.Len), true)
+	if err != nil {
+		return &pb.ReadStreamResponse{}, err
+	}
+
+	return &pb.ReadStreamResponse{
+		Data: data,
+	}, nil
+}
+
+func (a *agentGRPC) ReadStderr(ctx context.Context, req *pb.ReadStreamRequest) (*pb.ReadStreamResponse, error) {
+	data, err := a.pod.readStdio(req.ContainerId, req.ProcessId, int(req.Len), false)
+	if err != nil {
+		return &pb.ReadStreamResponse{}, err
+	}
+
+	return &pb.ReadStreamResponse{
+		Data: data,
+	}, nil
+}
+
+func (a *agentGRPC) CloseStdin(ctx context.Context, req *pb.CloseStdinRequest) (*gpb.Empty, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		file = proc.stdin
+	}
+
+	if err := file.Close(); err != nil {
+		return emptyResp, err
+	}
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) TtyWinResize(ctx context.Context, req *pb.TtyWinResizeRequest) (*gpb.Empty, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	if proc.termMaster == nil {
+		return emptyResp, fmt.Errorf("Terminal is not set, impossible to resize it")
+	}
+
+	fd := int(proc.termMaster.Fd())
+
+	// Get current terminal size.
+	winSize, err := unix.IoctlGetWinsize(fd, unix.TIOCGWINSZ)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	winSize.Row = uint16(req.Row)
+	winSize.Col = uint16(req.Column)
+
+	// Set new terminal size.
+	if err := unix.IoctlSetWinsize(fd, unix.TIOCSWINSZ, winSize); err != nil {
+		return emptyResp, err
+	}
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) StartSandbox(ctx context.Context, req *pb.StartSandboxRequest) (*gpb.Empty, error) {
+	if a.pod.running == true {
+		return emptyResp, fmt.Errorf("Pod already started, impossible to start again")
+	}
+
+	a.pod.id = req.Hostname
+	a.pod.network.dns = req.Dns
+	a.pod.running = true
+
+	if err := mountShareDir(req.Storage.Source, req.Storage.MountPoint, req.Storage.Fstype); err != nil {
+		return emptyResp, err
+	}
+
+	a.pod.shareDir = req.Storage.MountPoint
+
+	if err := setupDNS(a.pod.network.dns); err != nil {
+		return emptyResp, err
+	}
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) DestroySandbox(ctx context.Context, req *pb.DestroySandboxRequest) (*gpb.Empty, error) {
+	if a.pod.running == false {
+		agentLog.WithField("pod", a.pod.id).Info("Pod not started, this is a no-op")
+		return emptyResp, nil
+	}
+
+	a.pod.Lock()
+	for key, c := range a.pod.containers {
+		if err := c.removeContainer(); err != nil {
+			return emptyResp, err
```

Comment:
I want to make sure we can properly tear down a pod and its setup. This will avoid any issues related to unmounted directories, network configuration or other.
Also, I think we might want to support (for the future) a `RestartSandbox()`. In that case, we would save the launching time of the VM and it would result into a faster accessible pod.

Reply:
Fair enough. Just want to remind that if there are any issues with unmounted directories, network configuration or other, `DestroySandbox()` returns an error, and the only option runtime has is to shutdown the vm anyway.

BTW, it seem `c.removeContainer()` does not kill/stop the container but just wait endlessly. Then `DestroySandbox()` possibly becomes an API running forever?